test/service_test.cc

```diff
@@ -0,0 +1,458 @@
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <sys/time.h>
+#include <thread>
+#include <vector>
+
+#include <glog/logging.h>
+#include "gtest/gtest.h"
+
+#include "service_client_lib.h"
+#include "service_data_structure.h"
+
+namespace {
+
+uint64_t PrintId(const std::string &id) {
+  return *(reinterpret_cast<const uint64_t*>(id.c_str()));
+}
+
+const size_t kNumOfUsersPreset = 10;
+const size_t kNumOfUsersTotal = 20;
+const char *kShortText = "short";
+const char *kLongText = "longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong";
+
+// This test cases on the `ServiceDataStructure` to check whether their
+// interfaces work correctly.
+class ServiceTestDataStructure : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    chirp_connect_backend::backend_client_.reset(new BackendClientDebug());
+
+    // Set up users to be used in the sub-tests
+    for(size_t i = 0; i < kNumOfUsersTotal; ++i) {
+      user_list_.push_back(std::string("user") + std::to_string(i));
+      if (i < kNumOfUsersPreset) {
+        service_data_structure_.UserRegister(user_list_[i]);
+      }
+    }
+  }
+
+  // The container to store the usernames
+  std::vector<std::string> user_list_;
+
+  // The object that this test is going to test
+  ServiceDataStructure service_data_structure_;
+};
+
+// This tests on the `UserRegister` and `Login` in `ServiceDataStructure`
+TEST_F(ServiceTestDataStructure, UserRegisterAndLoginTest) {
```

Comment:
This test should be broken into 4 tests.
1. testing Register already register
2. testing Register new users
3. Login to registered users
4. Login to unregistered users